### PR TITLE
In Windows, attempt to use CreateProcess to launch [block|wallet|alert] notifications

### PR DESCRIPTION
**Pull request by dfletcher**
Motivation: In Windows, ::system() causes a terminal console window to pop up. This window grabs the focus away from the user, an irritating interruption to workflow.

Solution: Attempt to use CreateProcess() first. If this doesn't work for any reason, fall back to the previous solution and call ::system().

### CODE REVIEW

File: src/util.cpp [hunk 1]
```diff
@@ -682,7 +682,36 @@ boost::filesystem::path GetTempPath() {
...
+    // If we're in Windows, try using CreateProcess first.
+    ULONG nExitCode;
+    STARTUPINFO info={sizeof(info)};
+    ZeroMemory(&info, sizeof(STARTUPINFO));
+    info.cb = sizeof(STARTUPINFO);
+    info.dwFlags = STARTF_USESHOWWINDOW;
+    info.wShowWindow = SW_HIDE;
+    PROCESS_INFORMATION processInfo;
+    ZeroMemory(&processInfo, sizeof(PROCESS_INFORMATION));
+    if (CreateProcess(NULL, &strCommand[0], NULL, NULL, false, 0, NULL, NULL, &info, &processInfo))
```
Comment (laanwj) [inline src/util.cpp]: please use `strCommand.c_str()` (if you need zero-terminated) or `strCommand.data()` (if no zero terminated is necessary) i.s.o `&strCommand[0]`

File: src/util.cpp [hunk 2]
```diff
@@ -682,7 +682,36 @@ boost::filesystem::path GetTempPath() {
...
+        // CreateProcess succeeded so wait, check exit code and clean up.
+        WaitForSingleObject(processInfo.hProcess, INFINITE);
+        if(!GetExitCodeProcess(processInfo.hProcess, &nExitCode))
+          nExitCode = 1;
+        CloseHandle(processInfo.hThread);
+        CloseHandle(processInfo.hProcess);
+        nErr = (int)nExitCode; // Had to pass a ULONG to GetExitCodeProcess().
+    }
+    else
+    {
```
Comment (laanwj) [inline src/util.cpp]: Are you sure that this fallback is needed? I'm not sure what the implementation of mingw's system() does internally, but I'd assume it also makes use of CreateProcess or another win32 call.

**Comment by dfletcher** [comment]
Laanwj thanks for the feedback. Not quite sure why I did that &strCommand[0], I'll fix and reroll today. Regarding the second item, I believe system() handles more types of files, e.g. .bat files and document opening. I could be wrong about this. I will do some experimenting with both calls under MinGW and also possibly reroll that today.

**Comment by sipa** [comment]
Perhaps you can call ShellExecute, instead of CreateProcess?

**Comment by laanwj** [comment]
Yes, at least bat files would need to be supported. Document opening isn't that terribly important here, if one really wants to do that they could provide the program to launch it with in the command line.

**Comment by jgarzik** [comment]
Looks like a good start.  Needs more testing.

**Comment by dfletcher** [comment]
Okay finally had some time to research this a little better. Thanks everyone here for the suggestions.

I am not quite sure how to proceed, so here is what I have discovered, perhaps everyone here can make suggestions about the best route forward.

[Here's the gist of the program I used for testing](https://gist.github.com/dfletcher/e7ee74166ce9bd6bdd58).

|                          | CreateProcess() | ShellExecute() | ::system() |
| -------------------- |:--------------------:|:------------------:|:------------:|
| **.bat script**      | n s                     | n a                   | p s           |
| **self exe**         | n a                     | n a                   | p a           |
| **pythonw.exe**  | n s                     | n a                   | p s            |
| **python.exe**    | n s                     | n a                   | p s            |
<sup>**Key**: p = popup, n = no popup, s = synchronous, a = asynchronous</sup>

Each method has own strengths and weaknesses.

### CreateProcess()

**Pros**
* No popups.
* Straightforward command implementation, pass whole command string to windows.

**Cons**
* As laanwj pointed out it is using &strCommand[0]. Reason is that CreateProcessW might change that data, so we can't pass a const. "[The Unicode version of this function, CreateProcessW, can modify the contents of this string](https://msdn.microsoft.com/en-us/library/windows/desktop/ms682425%28v=vs.85%29.aspx)". Why this affects the non "W" version I have no idea. Is the fix to copy the string? Or to just cast c_str() to (char*)? Unsure.

### ShellExecute()

**Pros**
* Very consistent results while using SW_HIDE in the sixth argument of ShellExecute().

**Cons**
* Command must be parsed. The gist above only breaks on first whitespace, this would need to be smarter (accept a quoted string).

### ::system()
**Pros**
* Simple.

**Cons**
* Popups everywhere!

Thoughts and opinions? Thanks again for the feedback!

**Comment by laanwj** [comment]
Seems to me that CreateProcess comes out best. Especially if it can work without fallback to ::system.

> Why this affects the non "W" version I have no idea. Is the fix to copy the string? Or to just cast c_str() to (char*)? Unsure.

Just make a copy. Strings are not allowed to be mutated in-place in C++, but you could pass a begin_ptr(x) of a std::vector instead.

**Comment by dfletcher** [comment]
Ok updated. Nice call on begin_ptr() laanwj that makes it nice and clean. Since we removed the fallback wasn't quite sure what to do with nErr in that case, so just set it to -1 if CreateProcess() failed.

**Comment by kanzure** [comment]
Does the walletnotify hook need to be stress tested here? Sometimes there are quite a few created simultaneously.

**Comment by dfletcher** [comment]
@kanzure well it's anecdotal but I did just test 10 days of blocknotify with that last commit. Worked great, barely a blip on the CPU radar (admittedly fast 4 core modern machine). My test script this time is just a .bat piping the block hash to a file. It added 145 blocks to the output file really quickly and was not noticeably different from running it without the script. Would walletnotify be appreciably different from that? What's a good way to test it I wonder? Guess I could trash my blockchain and let it reload with a walletnotify script. Suggestions on what to put in it?

**Comment by dfletcher** [comment]
I'm confused why Travis failed here. Is that something in my fork? Did I screw up the merge from bitcoin/master?

**Comment by dfletcher** [comment]
Thanks a lot to jonasschnelli who helped me a ton in IRC figuring out how to work with github (this is my first pull request!)

**Comment by dfletcher** [comment]
@laanwj It seems that the fallback was necessary for the case of calling a terminal command like the "echo" in [this test case](https://github.com/bitcoin/bitcoin/blob/master/src/test/alert_tests.cpp#L159) (which is what fails w commit  01087f7 above).

I can imagine one other solution to this which I just tested with my gist ([updated](https://gist.github.com/dfletcher/e7ee74166ce9bd6bdd58)) locally. We can prepend `C:\Windows\System32\cmd.exe /C ` to the user command and force it to execute in a shell. Of course, the code would need to be better than just hard coding C:\Windows.

tl;dr: The two choices I can think of is either use the fallback solution or call an interpreter by hand.

Other ideas?

**Comment by dfletcher** [comment]
Still no opinions? Well it's easy for me to change this code to go either route, fallback or call cmd.exe. I would really like to have some solution for this even if it still pops up the window in some cases (the fallback solution).

**Comment by laanwj** [comment]
@dfletcher It's IMO not necessary to support providing a terminal command directly. As I said, what needs to be supported are .exe's and .bat's. The rest can be done from there.

As said, if remotely possible there should only be one path and no chain of fallbacks. If it fails, it fails.

**Comment by dfletcher** [comment]
@laanwj there is a test in the system which is testing using `echo`. This fails if we do not provide a fallback or use the cmd.exe interpreter directly. See the Travis failure above + two successes afterwards, failure uses no fallback or interpreter, first success is just putting the fallback back, second success is with an interpreter.

So that leaves (1) Use fallback. (2) Use an interpreter. (3) Remove [this test that uses "echo"](https://github.com/bitcoin/bitcoin/blob/master/src/test/alert_tests.cpp#L159) which is not an exe it's an interpreter command.

You seem to be suggesting #3 which doesn't sound good to me. Changing the test conditions because of a difference in the API calls we're making seems backwards to me.

**Comment by gavinandresen** [comment]
-alertnotify is a bitcoind, not Bitcoin-Qt, feature-- I wouldn't expect people running Bitcoin-Qt to run with -alertnotify, they're notified of alerts through the GUI.

So I think I'm NACK on this pull request in general for solving a problem that I don't think needs solving (command window popping up if you -alertnotify with Bitcoin-Qt).

**Comment by dfletcher** [comment]
@gavinandresen It solves -blocknotify and -walletnotify popping up a window also. -alertnotify is incidental.

File: src/util.cpp [hunk 3]
```diff
@@ -697,7 +697,43 @@ boost::filesystem::path GetTempPath() {
...
+    info.dwFlags = STARTF_USESHOWWINDOW;
+    info.wShowWindow = SW_HIDE;
+    PROCESS_INFORMATION processInfo;
+    ZeroMemory(&processInfo, sizeof(PROCESS_INFORMATION));
+    char system32[256];
+    if (GetSystemDirectory(system32, 255)) {
+        std::stringstream ssCommand;
+        ssCommand << system32 << "\\cmd.exe /C " << strCommand;
+        std::string strTempCmd = ssCommand.str();
+        std::vector<char> vecCommand(strTempCmd.begin(), strTempCmd.end() + 1);
```
Comment (laanwj) [inline src/util.cpp]: It is not allowed to use end()+1: end() is the end of the data. There is guarantee that there is anything after it, let alone a \0 byte.

This is becoming quite ugly, but I guess what you could do is (this gets rid of the stringstream too)
```
std::string strTempCmd = std::string(system32) + "\\cmd.exe /C " + strCommand;
const char *sTempCmd = ssCommand.c_str();
std::vector<char> vecCommand(sTempCmd, sTempCmd + strlen(sTempCmd) + 1);
```

File: src/util.cpp [hunk 4]
```diff
@@ -697,7 +697,43 @@ boost::filesystem::path GetTempPath() {
...
+        if (CreateProcess(NULL, begin_ptr(vecCommand), NULL, NULL, false, 0, NULL, NULL, &info, &processInfo))
```
Comment (laanwj) [inline src/util.cpp]: Code style: we have `{` on the same line as the if/else.

File: src/util.cpp (hunk 3 above)
Comment (laanwj) [inline src/util.cpp]: OTOH I don't see anywhere in the [CreateProcess](https://msdn.microsoft.com/en-us/library/windows/desktop/ms682425%28v=vs.85%29.aspx) documentation that it will actually try to change the passed-in string. So this should suffice:
```
std::string strTempCmd = std::string(system32) + "\\cmd.exe /C " + strCommand;
if (CreateProcess(NULL, const_cast<char*>(strTempCmd.c_str()), NULL, NULL, false, 0, NULL, NULL, &info, &processInfo))
```

**Comment by laanwj** [comment]
I do tend to agree with @gavinandresen that it'not s terribly useful to use `-*notify` with the GUI. For the daemon `bitcoind.exe` this is not an issue as it has its own shell window so should never pop up a new one?

**Comment by laanwj** [comment]
Closing due to inactivity. There are still some comments open. Let me know when you start working on this again, and I'll reopen.
